src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,187 @@
+            tail.next = newNode;
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index > size || index < 0) {
+            throw new IndexOutOfBoundsException("Index " + index + " is out of list");
+        }
+        if (size == index) {
+            add(value);
+        } else {
+            Node<T> currentNode = head;
+            while (nodeIndex != index) {
+                currentNode = currentNode.next;
+                nodeIndex++;
+            }
+            Node<T> addedNode = new Node<>(currentNode.prev, value, currentNode);
+            if (nodeIndex == 0) {
+                head.prev = addedNode;
+                addedNode.prev = null;
+                addedNode.next = head;
+                head = addedNode;
+            } else {
+                currentNode.prev.next = addedNode;
+                currentNode.prev = addedNode;
+                nodeIndex = 0;
+            }
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T listElement : list) {
+            Node<T> newNode = new Node<>(tail, listElement, null);
+            tail.next = newNode;
+            tail = newNode;
+            size++;
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> currentNode = head;
+        while (nodeIndex != index) {
+            currentNode = currentNode.next;
+            nodeIndex++;
+        }
+        nodeIndex = 0;
+        return currentNode.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        T oldNode = null;
+        Node<T> currentNode = head;
+        while (nodeIndex != index) {
+            currentNode = currentNode.next;
+            nodeIndex++;
+        }
+        oldNode = currentNode.value;
+        currentNode.value = value;
+        nodeIndex = 0;
+        return oldNode;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        T removedValue = null;
+        Node<T> currentNode = head;
+        if (index == 0 && size > 1) {
+            removedValue = currentNode.value;
+            currentNode.next.prev = null;
+            head = currentNode.next;
+            currentNode.next = null;
+        } else if (index == 0 && size == 1) {
+            currentNode.value = null;
+            currentNode.next = null;
+        } else {
+            while (nodeIndex != index) {
+                currentNode = currentNode.next;
+                nodeIndex++;
+            }
+            removedValue = currentNode.value;
+            currentNode.prev.next = currentNode.next;
+            if (size - index != 1) {
+                currentNode.next.prev = currentNode.prev;
+            }
+            nodeIndex = 0;
+        }
+        size--;
```

Comment:
lets use getNode(index) and unlink(node)

https://mate-academy.github.io/jv-program-common-mistakes/java-core/collections/linked-list.html#create-a-separate-method-unlinknode-node